test/formal/checked_uint_mul_12.py

```diff
@@ -25,13 +24,11 @@
 	# cast to full n_bits values
 	X = BVUnsignedUpCast(X_short, n_bits)
 	Y = BVUnsignedUpCast(Y_short, n_bits)
+	product_raw = MUL(X, Y)
+	#remove any overflown bits
+	product = BVUnsignedCleanupFunction(product_raw, type_bits)
 
-	# Constants
-	maxValue = BVUnsignedMax(type_bits, n_bits)
-
-	# Overflow check in YulUtilFunction::overflowCheckedIntMulFunction
-	overflow_check = AND(ISZERO(ISZERO(X)), GT(Y, DIV(maxValue, X)))
+	# Overflow check in YulUtilFunction::overflowCheckedIntMulFunctions
+	overflow_check = AND(ISZERO(ISZERO(X)), ISZERO(EQ(Y, DIV(product, X))))
 
 	rule.check(overflow_check != 0, actual_overflow)
```

Comment:
Also note that this works as well:

```suggestion
	if type_bits <= n_bits / 2:
		overflow_check = ISZERO(EQ(product, product_raw))
		rule.check(overflow_check != 0, actual_overflow)
	else:
		# Overflow check in YulUtilFunction::overflowCheckedIntMulFunctions
		overflow_check = ISZERO(OR(ISZERO(X), EQ(Y, DIV(product, X))))

		rule.check(overflow_check != 0, actual_overflow)
```
(with which we should also change the range of ``type_bits`` to ``type_bits in [4, 6, 8, 12]`` to check the borderline case)


So I think we should add yet another case distinction to the unsigned checked multiplication function. For all types up to 128 bits, we can just do

```
 				let product_raw := mul(x, y)
 				product := <cleanupFunction>(product_raw)
 				if iszero(eq(product, product_raw)) { <panic>() }
```

This is bound to be cheaper by quite some margin.

Reply:
The same is also true for the signed case.

Reply:
Yeah, that should be cheaper. I will implement the changes in the code and post the benchmarks after.